### DISCUSSION TITLE
Best configuration to avoid Deadlock in SELECT .. FOR UPDATE?

### DISCUSSION THREAD

**Discussion by garywhiteford** [apache/airflow#22796] [Q&A]
https://github.com/apache/airflow/blob/4b25cb9d08565502172cb847c79d81559775d504/airflow/jobs/scheduler_job.py#L1492-L1499

My client is using Airflow v1.10.9. Every once in a while they hit a Deadlock when this query is run by the scheduler and tasks (even if successful) are not marked correctly. 

Is there a configuration in Airflow v1.10.9 that I can change to avoid this? 

Alternatively (and I realize I am creeping outside scope), is there a configuration in the MySQL database I am using that needs to be made? 

Thanks in advance for any help! 

P.S. I *am* encouraging my customer to upgrade. 

### Error in Scheduler log
```
93634:Apr 02 18:11:00 [schedulerHost] bash[2900]: [2022-04-02 18:11:00,893] {scheduler_job.py:1497} ERROR - Error queuing tasks
93635-Apr 02 18:11:00 [schedulerHost] bash[2900]: [2022-04-02 18:11:00,894] {scheduler_job.py:1498} ERROR - (_mysql_exceptions.OperationalError) (1213, 'Deadlock found when trying to get lock; try restarting transaction')
93636:Apr 02 18:11:00 [schedulerHost] bash[2900]: sqlalchemy.exc.OperationalError: (_mysql_exceptions.OperationalError) (1213, 'Deadlock found when trying to get lock; try restarting transaction')
93637-Apr 02 18:11:00 [schedulerHost] bash[2900]: [SQL: SELECT [bunch-o-fields]
93638-Apr 02 18:11:00 [schedulerHost] bash[2900]: FROM task_instance LEFT OUTER JOIN dag_run ON task_instance.dag_id = dag_run.dag_id AND task_instance.execution_date = dag_run.execution_date
93639-Apr 02 18:11:00 [schedulerHost] bash[2900]: WHERE task_instance.dag_id IN ([100+ %s, placeholders]) AND task_instance.state IN (%s, %s, %s) AND (dag_run.state != %s OR dag_run.state IS NULL) FOR UPDATE]
93640-Apr 02 18:11:00 [schedulerHost] bash[2900]: [parameters: ([100+ dag IDs], 'queued', 'scheduled', 'up_for_reschedule', 'running')]
```

### Configurations in the environment that might apply

```
export AIRFLOW__CORE__SQL_ALCHEMY_POOL_RECYCLE=300
export AIRFLOW__CORE__SQL_ALCHEMY_POOL_PRE_PING=True
export AIRFLOW__CORE__SQL_ALCHEMY_RECONNECT_TIMEOUT=300
export AIRFLOW__CORE__PARALLELISM=300
export AIRFLOW__CORE__DAG_CONCURRENCY=300
```

**Comment by potiuk** [comment]
Since Airflow 1.10 is end of life as of June 2021 and will never receive any fixes (not even critical security ones), and 1.10.9 has been released more than 2 years ago, the best configuration (by far) is to upgrade your client to Airflow 2.

Since airflow entirely 1.10 is end-of-life elready quite a long time you are unlikely to get any support here.

P.S. Tell your customer that they open themselves to Log4shell or Spring4Shell vulnerabilties if they won't upgrade. There was not a single upgrade to 1.10.9 deps for more than 2 years. And there were already security fixes (you can look them up) that were only implemented in Airflow 2. 

Just saying.

**Reply by garywhiteford** [reply]
Understood!
